jetty-core/jetty-websocket/jetty-websocket-jetty-api/src/main/java/org/eclipse/jetty/websocket/api/Frame.java

```diff
@@ -101,4 +101,126 @@ public String toString()
     boolean isRsv2();
 
     boolean isRsv3();
+
+    /**
+     * The effective opcode of the frame accounting for the CONTINUATION opcode.
+     * If the frame is a CONTINUATION frame for a TEXT message, this will return TEXT.
+     * If the frame is a CONTINUATION frame for a BINARY message, this will return BINARY.
+     * Otherwise, this will return the same opcode as the frame.
+     * @return the effective opcode of the frame.
+     */
+    byte getEffectiveOpCode();
+
+    class Wrapper implements Frame
+    {
+        private final Frame _frame;
+
+        public Wrapper(Frame frame)
+        {
+            _frame = frame;
+        }
+
+        @Override
+        public byte[] getMask()
+        {
+            return _frame.getMask();
+        }
+
+        @Override
+        public byte getOpCode()
+        {
+            return _frame.getOpCode();
+        }
+
+        @Override
+        public ByteBuffer getPayload()
+        {
+            return _frame.getPayload();
+        }
+
+        @Override
+        public int getPayloadLength()
+        {
+            return _frame.getPayloadLength();
+        }
+
+        @Override
+        public Type getType()
+        {
+            return _frame.getType();
+        }
+
+        @Override
+        public boolean hasPayload()
+        {
+            return _frame.hasPayload();
+        }
+
+        @Override
+        public boolean isFin()
+        {
+            return _frame.isFin();
+        }
+
+        @Override
+        public boolean isMasked()
+        {
+            return _frame.isMasked();
+        }
+
+        @Override
+        public boolean isRsv1()
+        {
+            return _frame.isRsv1();
+        }
+
+        @Override
+        public boolean isRsv2()
+        {
+            return _frame.isRsv2();
+        }
+
+        @Override
+        public boolean isRsv3()
+        {
+            return _frame.isRsv3();
+        }
+
+        @Override
+        public byte getEffectiveOpCode()
+        {
+            return _frame.getEffectiveOpCode();
+        }
+    }
+
+    static Frame copy(Frame frame)
+    {
+        ByteBuffer payloadCopy = copy(frame.getPayload());
+        return new Frame.Wrapper(frame)
+        {
+            @Override
+            public ByteBuffer getPayload()
+            {
+                return payloadCopy;
+            }
+
+            @Override
+            public int getPayloadLength()
+            {
+                return payloadCopy == null ? 0 : payloadCopy.remaining();
+            }
+        };
+    }
+
+    private static ByteBuffer copy(ByteBuffer buffer)
```

Comment:
use `BufferUtil.copy`?

Reply:
I don't think we want to introduce a dependency from the jetty-ws-api to jetty-util just for this.